# Imports

In [1]:
import multiprocessing
import random
# import string
import time
import threading

# Functions

In [2]:
def generate_and_add_numbers(n: int = 1000):
    total = 0
    for i in range(n):
        total += random.randint(0,1000000)
    return total


def generate_and_join_letters(n: int = 1000):
    letters = ''
    for i in range(n):
        letters += chr(random.randint(33, 126))
        # letters += random.choice(string.ascii_letters + string.digits)
    return letters

# Seauential Execution

In [3]:
print("Starting the Program")
total_start_time = time.time()

generate_and_add_numbers(int(1e7))
generate_and_join_letters(int(1e7))

total_end_time = time.time()
print("Exiting the Program")
sequential_execution_time = total_end_time - total_start_time
print(f"It took {sequential_execution_time}s to execute the tasks.")

Starting the Program


Exiting the Program
It took 7.206875801086426s to execute the tasks.


# Threads execution

In [4]:
print("Starting the  thread program")
total_start_time = time.time()

thread_numbers = threading.Thread(target=generate_and_add_numbers, args=[int(1e7)])
thread_letters = threading.Thread(target=generate_and_join_letters, args=[int(1e7)])

thread_numbers.start()
thread_letters.start()

thread_numbers.join()
thread_letters.join()

total_end_time = time.time()
print("Exiting the thread program")
thread_execution_time = total_end_time - total_start_time
print(f"It took {thread_execution_time}s to execute the tasks with thread.")

Starting the  thread program
Exiting the thread program
It took 7.118330478668213s to execute the tasks with thread.


# Trials with the same threads

In [5]:
print("Starting the two threads program for generate numbers")
total_start_time = time.time()

thread_numbers = threading.Thread(target=generate_and_add_numbers, args=[int(1e7)])
thread_letters = threading.Thread(target=generate_and_add_numbers, args=[int(1e7)])

thread_numbers.start()
thread_letters.start()

thread_numbers.join()
thread_letters.join()

total_end_time = time.time()
print("Exiting two threads program for generate numbers")
execution_time = total_end_time - total_start_time
print(f"It took {execution_time}s to execute the tasks with threads.")

Starting the two threads program for generate numbers


Exiting two threads program for generate numbers
It took 6.730978012084961s to execute the tasks with threads.


In [6]:
print("Starting the two threads program for generate letters")
total_start_time = time.time()

thread_numbers = threading.Thread(target=generate_and_add_numbers, args=[int(1e7)])
thread_letters = threading.Thread(target=generate_and_add_numbers, args=[int(1e7)])

thread_numbers.start()
thread_letters.start()

thread_numbers.join()
thread_letters.join()

total_end_time = time.time()
print("Exiting two threads program for generate letters")
execution_time = total_end_time - total_start_time
print(f"It took {execution_time}s to execute the tasks with threads.")

Starting the two threads program for generate letters
Exiting two threads program for generate letters
It took 6.020277738571167s to execute the tasks with threads.


# Trials With processes

In [16]:
print("Starting the two processs program for generate letters")
total_start_time = time.time()

process_numbers = multiprocessing.Process(target=generate_and_add_numbers, args=[int(1e7)])
process_letters = multiprocessing.Process(target=generate_and_add_numbers, args=[int(1e7)])

process_numbers.start()
process_letters.start()

process_numbers.join()
process_letters.join()

total_end_time = time.time()
print("Exiting two processs program for generate letters")
process_execution_time = total_end_time - total_start_time
print(f"It took {process_execution_time}s to execute the tasks with processs.")

Starting the two processs program for generate letters
Exiting two processs program for generate letters
It took 2.958604574203491s to execute the tasks with processs.


In [18]:
serial = 0.005/2.95

# Add your interpretations and conclusions here
- Computing the speedups

In [8]:
speedup_thread = sequential_execution_time/thread_execution_time
speedup_processes = sequential_execution_time/process_execution_time

print(f"The speedup using threads is {speedup_thread}")
print(f"The speedup using processes is {speedup_processes}")

The speedup using threads is 1.0124390575407478
The speedup using processes is 2.2849227500019653


- Computing the Efficiency 

In [20]:
np = 4
efficiency_thread = speedup_thread/np
efficiency_processes = speedup_processes/np

print(f"The efficiency using threads is {efficiency_thread}")
print(f"The efficiency using processes is {efficiency_processes}")

The efficiency using threads is 0.25310976438518695
The efficiency using processes is 0.5712306875004913


- Amdhal and Gustafson Laws

In [19]:
P = 0.99
amdhal_s=1/(1-P)
print(amdhal_s)

99.99999999999991


In [25]:
alpha = 1-0.99
gustafson= np - alpha*(np-1)
print(gustafson)

3.9699999999999998


# conclusion

#### **Threads vs. Processes – Which Worked Better?**  
- Using **threads** didn’t help much (**7.12s vs. 7.21s sequential**) because of Python **GIL**, which limits parallel execution.  
- Adding **two threads** improved things slightly (**6.73s for numbers, 6.02s for letters**), but not significantly.  
- **Multiprocessing**, on the other hand, was much faster (**2.96s**), showing a clear advantage for CPU-heavy tasks.  

#### **Speedup & Efficiency – Was Parallelization Worth It?**  
- **Threads** gave a tiny speed boost (**1.01x faster**), with an efficiency of **25.3%**—not great.  
- **Processes** were **2.28x faster** and had a much better efficiency of **57.1%**.  

#### **What Does This Mean?**  
- **Threads aren’t useful** for tasks that rely heavily on the CPU, but they’re good for I/O-bound tasks.  
- **Multiprocessing is the way to go** for CPU-heavy tasks since it takes full advantage of multiple cores.  
- According to **Gustafson’s Law**, if we scale this up, we could see even bigger speed improvements (**~3.97x speedup**).  

**Bottom line:** If you’re working with CPU-intensive tasks, multiprocessing is the best!